### ISSUE TITLE
User implemented mapping method is not used on nullable Enum

### ISSUE THREAD

**Issue by Valkozaur** [riok/mapperly#837]
**Describe the bug**
When you have a user implemented Enum mapping method the mapper the method is only used if the return type of the method is not marked as Nullable. 

**To Reproduce**
1. Declare a mapper
2. Add a mapping method `CarDto CarToDto(Car car);
3. Have a nullable Enum property in the CarDto.
4. Create user implemented method for the Enum mapping returning a nullable Enum value
5. The user implemented mapping is not used in the generated file.
6. If you create the user implemented mapping as a non-nullable return type it is used.

**Expected behavior**
That the nullable Enum method is going to be used when the property is marked as a Nullable Enum. 

**Code snippets**
```c#
public enum EngineType
{
    Electric = 1,
    Petrol = 2
}

public class Car
{
    public string Engine { get; set; }
}

public class CarDto
{
    public EngineType? Engine { get; set; }
}


[Mapper]
public static partial class CarMapper
{
    public static partial CarDto ToCarDto(this Car car);

    //This is picked as an user implemented mapping
    private static EngineType MapEngine(string engineType) => engineType[0] switch
    {
        'E' => EngineType.Electric,
        'P' => EngineType.Petrol
    };

    //This is not used as an user implemented mapping
    private static EngineType? MapNullEngine(string engineType) => engineType[0] switch
    {
        'E' => EngineType.Electric,
        'P' => EngineType.Petrol,
        _ => null
    };
}
```
 

**Environment (please complete the following information):**
- Mapperly Version: 3.2.0
- .NET Version: NET7
- Target Framework: net7
- IDE: Visual Studio 2022 Preview 17.8.0
- OS: Windows

**Comment by TimothyMakkison** [comment]
Hey, thanks for creating this issue. Your example is broken, what happens when you use my code below? It doesn't look like my generated code is incorrect.

## My recreation
```C#
public enum EngineType { Electric = 1, Petrol = 2 }

public class Car { public string Engine { get; set; } }

public class CarDto { public EngineType? Engine { get; set; } }

[Mapper]
public static partial class CarMapper
{
    public static partial CarDto ToCarDto(this Car car);

    //This is picked as an user implemented mapping
    private static EngineType MapEngine(string engineType) => engineType switch
    {
        "E" => EngineType.Electric,
        "P" => EngineType.Petrol
    };

    //This is not used as an user implemented mapping
    private static EngineType? MapNullEngine(string engineType) => engineType switch
    {
        "E" => EngineType.Electric,
        "P" => EngineType.Petrol,
        _ => null
    };
}
```

## Generated code
```C#
// <auto-generated />
#nullable enable
public static partial class CarMapper
{
    public static partial global::CarDto ToCarDto(this global::Car car)
    {
        var target = new global::CarDto();
        target.Engine = MapNullEngine(car.Engine);
        return target;
    }
}
```

**Comment by Valkozaur** [comment]
Indeed your code runs correctly. Now I am not sure why but in my case it does not work. I will try my best to re-create the problem and come back to you. 

Best regards, Valentin!

**Comment by latonz** [comment]
@Valkozaur  in your code the user implemented mapping methods don't have method names which is not valid c# and there is also a wrong brace. Not sure if these are just copy/paste or simplifying errors....

**Comment by Valkozaur** [comment]
Thanks for the feedback, I tried writing the code myself in the issue description. It's not a copy paste of the original problem as it occurs in my company's private code. Currently I am trying to pinpoint the problem, it's a fact that @TimothyMakkison code works even copy/pasted in the same solution. So it must be something on my side, probably connected with the DTOs. 

The source is a generated GRPC model, using Google.Protobufs. It might be the reason. Till I find the problem, please close the issue.

**Comment by TimothyMakkison** [comment]
Good luck, perhaps some of your DTOs are in a nullable disabled context?

**Comment by Valkozaur** [comment]
Hey @TimothyMakkison nice guess. Indeed my source seems to be in a nullable disabled context. Unfortunately they are coming from a third party library and I do not have access to change the nullable context. Why is the source nullable context important, any idea how I can skip this check?

**Comment by TimothyMakkison** [comment]
Can't experiment rn but what happens if you change your user implemented mapping to accept a nullable input?

**Comment by Valkozaur** [comment]
You can achieve that with the same experiment as above. Just move the source to a new file, it will than pick the Non-nullable user implemented method.

```c#
#nullable disable
public class Car
{
    private string _engine;

    public string Engine { get => _engine; set => _engine = value; }
}
```

**Comment by Valkozaur** [comment]
Actually a work around is indeed to change the user implemented mapping to accept a nullable input. Though I am not 100% if this must be the expected behavior, at least it was not expected by me.

**Comment by latonz** [comment]
This is currently expected behaviour: a user implemented mapping is only used by Mapperly if the types do match exactly (including nullability).

**Comment by TimothyMakkison** [comment]
I haven't confirmed this but I'd argue this is a bug `Nullable -> Nullable` uses the method `None->None` (even though the mapping is now `None->Nullable`) after using a null guard. `None->Nullable` will use the method `None->Nullable`. By changing the source nullabiliy a different user method is used

**Comment by latonz** [comment]
This should be fixed in 3.3.0.
